src/solvers/refinement/string_refinement.h

```diff
@@ -28,6 +28,41 @@ Author: Alberto Griggio, alberto.griggio@gmail.com
 #include <solvers/refinement/string_refinement_invariant.h>
 
 #define MAX_NB_REFINEMENT 100
+#define CHARACTER_FOR_UNKNOWN '?'
+
+/// Similar interface to union-find for expressions, with a function for
+/// replacing sub-expressions by their result for find.
+class union_find_replacet
+{
+public:
+  /// return true if unmodified
+  bool replace_expr(exprt &expr) const
+  {
+    bool unchanged=::replace_expr(map_, expr);
+    while(!unchanged && !::replace_expr(map_, expr))
+      continue;
+    return unchanged;
+  }
+
+  exprt find_expr(exprt expr) const
+  {
+    replace_expr(expr);
+    return expr;
+  }
+
+  exprt add_symbol(const exprt &expr, const exprt &dst);
+
+  std::vector<std::pair<exprt, exprt>> to_vector() const
+  {
+    std::vector<std::pair<exprt, exprt>> equations;
+    for(const auto &pair : map_)
+      equations.emplace_back(pair.first, find_expr(pair.second));
+    return equations;
+  }
+
+private:
+  replace_mapt map_;
```

Comment:
Trailing `_` is unusual -- how about simply `replace_map`

Reply:
Disagree with this one. Having a special naming convention for data members helps to distinguish them from global data or arguments, without ugly names leaking to the public interface. Trailing underscores are better than leading underscores, which are reserved under some circumstances.

Reply:
We've been talking to @peterschrammel. He tolerates the idea of postfix underscore in data member names.